crates/wasmi/src/memory/buffer.rs

```diff
@@ -20,33 +27,157 @@ impl ByteBuffer {
     pub fn data(&self) -> &[u8] {
-        &self.bytes[..]
+        // Safety: either is backed by a `Vec` or a static buffer, ptr[0..len] is valid.
+        unsafe { core::slice::from_raw_parts(self.ptr, self.len) }
     }
 
     /// Returns an exclusive slice to the bytes underlying to the byte buffer.
     pub fn data_mut(&mut self) -> &mut [u8] {
-        &mut self.bytes[..]
+        // Safety: either is backed by a `Vec` or a static buffer, ptr[0..len] is valid.
+        unsafe { core::slice::from_raw_parts_mut(self.ptr, self.len) }
+    }
+}
+
+impl Drop for ByteBuffer {
+    fn drop(&mut self) {
+        if !self.is_static {
+            // Safety: those parts have been obtained from `Vec`.
+            unsafe { Vec::from_raw_parts(self.ptr, self.len, self.capacity) };
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    #[test]
+    fn test_basic_allocation_deallocation() {
+        let buffer = ByteBuffer::new(10);
+        assert_eq!(buffer.len(), 10);
+        // Dropping the buffer should not cause UB.
+    }
+
+    #[test]
+    fn test_basic_data_manipulation() {
+        let mut buffer = ByteBuffer::new(10);
+        assert_eq!(buffer.len(), 10);
+        let data = buffer.data(); // test we can read the data
+        assert_eq!(data, &[0; 10]);
+        let data = buffer.data_mut(); // test we can take a mutable reference to the data
+        data[4] = 4; // test we can write to the data and it is not UB
+        let data = buffer.data(); // test we can take a new reference to the data
+        assert_eq!(data, &[0, 0, 0, 0, 4, 0, 0, 0, 0, 0]); // test we can read the data
+                                                           // test drop is okay
+    }
+
+    #[test]
+    fn test_static_buffer_initialization() {
+        static mut BUF: [u8; 10] = [0; 10];
+        let buf = unsafe { &mut *core::ptr::addr_of_mut!(BUF) };
+        let mut buffer = ByteBuffer::new_static(buf, 5);
+        assert_eq!(buffer.len(), 5);
+        // Modifying the static buffer through ByteBuffer and checking its content.
+        let data = buffer.data_mut();
+        data[0] = 1;
+        unsafe {
+            assert_eq!(BUF[0], 1);
+        }
+    }
+
+    #[test]
+    fn test_growing_buffer() {
+        let mut buffer = ByteBuffer::new(5);
+        buffer.grow(10);
+        assert_eq!(buffer.len(), 10);
+        assert_eq!(buffer.data(), &[0; 10]);
+    }
+
+    #[test]
+    fn test_growing_static() {
+        static mut BUF: [u8; 10] = [0; 10];
+        let buf = unsafe { &mut *core::ptr::addr_of_mut!(BUF) };
+        let mut buffer = ByteBuffer::new_static(buf, 5);
+        assert_eq!(buffer.len(), 5);
+        assert_eq!(buffer.data(), &[0; 5]);
+        buffer.grow(8);
+        assert_eq!(buffer.len(), 8);
+        assert_eq!(buffer.data(), &[0; 8]);
+        buffer.grow(10);
+        assert_eq!(buffer.len(), 10);
+        assert_eq!(buffer.data(), &[0; 10]);
+    }
```

Comment:
Grow tests should probably have a non-zero static buffer and check if growing zeros the correct regions.